### ISSUE TITLE
Netty streams should reduce the capacity of pooled buffers passed to sendFrame

### ISSUE THREAD

**Issue by louiscryan** [grpc/grpc-java#452]
When a small write and flush is passed through the framer we hold the full 4k of the buffer until the write completes. By reducing the capacity of a pool direct buffer to the readably byte limit we immediately release the unwritten portion of the buffer back to the pool.

This may or may not have an impact on performance and utility should be evaluated by bencmarking. It may improve performance by making more bytes available to thread-local allocation. The most likely benchmarks to be impacted would be streaming ones that write and flush many messages in a tight loop as it would alleviate buffer-arena locks.

**Comment by ejona86** [comment]
@louiscryan, I don't quite follow what this issue is about. Is it to reduce the minimum buffer size permitted by [NettyWritableBufferAllocator](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java#L52).

**Comment by louiscryan** [comment]
No, it's to free th unused portion of a buffer that is being flushed back to the allocator. Take a look at

**Comment by ejona86** [comment]
I don't see the code that does what you are describing. I see code that re-assigns the capacity in-place if it makes sense (like the new capacity is not < 1/2 the true capacity), but otherwise [allocates](https://github.com/netty/netty/blob/master/buffer/src/main/java/io/netty/buffer/PooledByteBuf.java#L112) [a new](https://github.com/netty/netty/blob/833b92a5aaca287f0f9ebdc74d15c6f0a5baef74/buffer/src/main/java/io/netty/buffer/PoolArena.java#L284) [buffer](https://github.com/netty/netty/blob/833b92a5aaca287f0f9ebdc74d15c6f0a5baef74/buffer/src/main/java/io/netty/buffer/PoolThreadCache.java#L401) and [copies the contents](https://github.com/netty/netty/blob/833b92a5aaca287f0f9ebdc74d15c6f0a5baef74/buffer/src/main/java/io/netty/buffer/PoolArena.java#L294).

It never releases just the unused portion; it either releases the entire thing and makes a new one (and copies) or releases nothing (so should have no impact for us).
